### PR TITLE
Only try to connect ws immediately when tab is activated

### PR DESCRIPTION
**Pull request by mustard-mh**
## Description

- [ ] hold to remove debug code

## Related Issue(s)

Relates ENT-24

## How to test

‼️  Test with Firefox

**Regression**
- Workspace should work like before
- Integration tests should pass

**Firefox WS**
- Create a workspace in preview env with Firefox + VS Code Browser
- Open workspace with 20 more tabs
- Check if the tab you're working works well
- Wait for x minutes, and click other tabs, they should be activated after tab is activated

**Feature Flag**
Feature is controlled by ff `websocket_url_provider_returns_immediately`. Smoke test if workspace is working (need refresh web)

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - hw-ws</li>
	<li><b>🔗 URL</b> - <a href="https://hw-ws.preview.gitpod-dev.com/workspaces" target="_blank">hw-ws.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - hw-ws-gha.26818</li>
	<li><b>🗒️ Logs</b> - <a href="https://console.cloud.google.com/logs/query;query=jsonPayload.kubernetes.host%3D%22preview-hw-ws%22%0A%0A--%20Filter%20on%20service:%0A--%20jsonPayload.serviceContext.service%3D%22ws-manager-mk2%22%0A;duration=P1D?project=gitpod-dev-preview" target="_blank">GCP Logs Explorer</a></li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [x] analytics=log
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [x] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [x] /werft preemptible
      Saves cost. Untick this only if you're really sure you need a non-preemtible machine.
- [x] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

**Comment by akosyakov** [comment]
it looks complicated, why do we do it?

**Comment by mustard-mh** [comment]
> it looks complicated, why do we do it?

@akosyakov  Firefox will block websocket if there're too many tabs opened, you could get more context from ENT-24 and [[internal chat](https://gitpod.slack.com/archives/C071G5TTS49/p1718957991592659)]

**Comment by akosyakov** [comment]
Is there a way we auto close tabs of old workspaces or something like that, I afraid that we will break something with all smartness around events.

**Comment by geropl** [comment]
> Is there a way we auto close tabs of old workspaces or something like that, I afraid that we will break something with all smartness around events.
> I remember we would show a notification to the user from the dashboard that failed to establish the websocket maybe you have so many tabs. But it was disabled for IDE, maybe we enable it in the dashboard always in case of FF.

They way I understand your comment is:
 - if a tab is in background for long enough, we stop re-connecting websocket, and show a "Old tab, want to reconnect? Ok" modal
 - if users come back, they have to manually click that button, which:
   - makes them aware something might be off
   - we do a soft-reload in the background, mitigating any potential "state-out-of-sync" issues

@akosyakov Is that what you mean? :thinking:

File: components/dashboard/src/service/service.tsx [hunk 1]
```diff
@@ -28,18 +29,40 @@ import {
...
+import { getFeatureFlagValue } from "../experiments/flags";
+import UAParser from "ua-parser-js";
 
 export const gitpodHostUrl = new GitpodHostUrl(window.location.toString());
 
-function createGitpodService<C extends GitpodClient, S extends GitpodServer>() {
-    let host = gitpodHostUrl.asWebsocket().with({ pathname: GitpodServerPath }).withApi();
+const gitpodHost = gitpodHostUrl.asWebsocket().with({ pathname: GitpodServerPath }).withApi().toString();
 
+const isFirefox = new UAParser().getBrowser().name?.toLowerCase().includes("firefox") ?? false;
```
Comment (akosyakov) [inline components/dashboard/src/service/service.tsx:40]: fyi https://github.com/microsoft/vscode/blob/2e3a1d8bd9270480fc548d9351c7efebfd52bfed/src/vs/base/browser/browser.ts#L101-L103

Comment (mustard-mh) [reply components/dashboard/src/service/service.tsx:40]: The way we're using now is a copy from browser extension advertisement https://github.com/gitpod-io/gitpod/blob/bf27597ae5f69345ddb8a68727b74b6ac7c7c362/components/dashboard/src/workspaces/BrowserExtensionBanner.tsx#L109-L110

**Comment by mustard-mh** [comment]
Canceled, let's re-open if needed in the future

**Pull request closed without merging**